### PR TITLE
Refactor peep set name actions

### PR DESCRIPTION
**Pull request by oli414**
Refactors `GAME_COMMAND_SET_GUEST_NAME` and `GAME_COMMAND_SET_STAFF_NAME` for https://github.com/OpenRCT2/OpenRCT2/pull/5053

### CODE REVIEW

File: src/openrct2/actions/GameActionCompat.cpp [hunk 1]
```diff
@@ -201,4 +203,36 @@ extern "C"
     }
 #pragma endregion
 
+#pragma region GuestSetName
+
+    void guest_set_name(uint16 spriteIndex, const char *name)
```
Comment (janisozaur) [inline src/openrct2/actions/GameActionCompat.cpp]: asterisk should have spaces on both sides

File: src/openrct2/actions/GameActionCompat.cpp [hunk 2]
```diff
@@ -201,4 +203,36 @@ extern "C"
...
+    {
+        auto gameAction = GuestSetNameAction(spriteIndex, name);
+        GameActions::Execute(&gameAction);
+    }
+
+    /**
+    *
+    *  rct2: 0x00698D6C
+    */
+    void game_command_set_guest_name(sint32 *eax, sint32 *ebx, sint32 *ecx, sint32 *edx, sint32 *esi, sint32 *edi, sint32 *ebp) {
```
Comment (janisozaur) [inline src/openrct2/actions/GameActionCompat.cpp]: brace on new line, spaces after asterisks

File: src/openrct2/windows/Guest.cpp [hunk 3]
```diff
@@ -1147,9 +1147,8 @@ void window_guest_overview_text_input(rct_window *w, rct_widgetindex widgetIndex
         return;
 
     gGameCommandErrorTitle = STR_CANT_NAME_GUEST;
-    game_do_command(1, GAME_COMMAND_FLAG_APPLY, w->number, *((sint32*)(text + 0)), GAME_COMMAND_SET_GUEST_NAME, *((sint32*)(text + 8)), *((sint32*)(text + 4)));
-    game_do_command(2, GAME_COMMAND_FLAG_APPLY, w->number, *((sint32*)(text + 12)), GAME_COMMAND_SET_GUEST_NAME, *((sint32*)(text + 20)), *((sint32*)(text + 16)));
-    game_do_command(0, GAME_COMMAND_FLAG_APPLY, w->number, *((sint32*)(text + 24)), GAME_COMMAND_SET_GUEST_NAME, *((sint32*)(text + 32)), *((sint32*)(text + 28)));
+
+	guest_set_name(w->number, text);
```
Comment (janisozaur) [inline src/openrct2/windows/Guest.cpp]: misaligned

File: src/openrct2/windows/Staff.cpp [hunk 4]
```diff
@@ -1193,9 +1193,8 @@ void window_staff_overview_text_input(rct_window *w, rct_widgetindex widgetIndex
         return;
 
     gGameCommandErrorTitle = STR_STAFF_ERROR_CANT_NAME_STAFF_MEMBER;
-    game_do_command(1, GAME_COMMAND_FLAG_APPLY, w->number, *((sint32*)(text + 0)), GAME_COMMAND_SET_STAFF_NAME, *((sint32*)(text + 8)), *((sint32*)(text + 4)));
-    game_do_command(2, GAME_COMMAND_FLAG_APPLY, w->number, *((sint32*)(text + 12)), GAME_COMMAND_SET_STAFF_NAME, *((sint32*)(text + 20)), *((sint32*)(text + 16)));
-    game_do_command(0, GAME_COMMAND_FLAG_APPLY, w->number, *((sint32*)(text + 24)), GAME_COMMAND_SET_STAFF_NAME, *((sint32*)(text + 32)), *((sint32*)(text + 28)));
+
+	staff_set_name(w->number, text);
```
Comment (janisozaur) [inline src/openrct2/windows/Staff.cpp]: use spaces, not tabs

File: src/openrct2/actions/GuestSetNameAction.hpp [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        {
+            log_warning("Invalid game command for sprite %u", _spriteIndex);
+            return std::make_unique<GameActionResult>(GA_ERROR::INVALID_PARAMETERS, STR_NONE);
+        }
+
+        rct_string_id newUserStringId = user_string_allocate(USER_STRING_HIGH_ID_NUMBER | USER_STRING_DUPLICATION_PERMITTED, _name.c_str());
+        if (newUserStringId == 0)
+        {
+            // TODO: Probably exhausted, introduce new error.
+            return std::make_unique<GameActionResult>(GA_ERROR::UNKNOWN, STR_NONE);
```
Comment (janisozaur) [inline src/openrct2/actions/GuestSetNameAction.hpp]: Looks like this is communicated via `gGameCommandErrorText` can you set the error to have value of that variable?

File: src/openrct2/actions/GuestSetNameAction.hpp [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        }
+        user_string_free(newUserStringId);
+
+        return std::make_unique<GameActionResult>();
+    }
+
+    GameActionResult::Ptr Execute() const override
+    {
+        rct_string_id newUserStringId = user_string_allocate(USER_STRING_HIGH_ID_NUMBER | USER_STRING_DUPLICATION_PERMITTED, _name.c_str());
```
Comment (janisozaur) [inline src/openrct2/actions/GuestSetNameAction.hpp]: This still needs to be checked if valid strings was returned

File: src/openrct2/actions/GuestSetNameAction.hpp [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+        rct_peep *peep = GET_PEEP(_spriteIndex);
+        if (peep->type != PEEP_TYPE_GUEST)
+        {
+            log_warning("Invalid game command for sprite %u", _spriteIndex);
+            return std::make_unique<GameActionResult>(GA_ERROR::INVALID_PARAMETERS, STR_NONE);
+        }
+
+        set_format_arg(0, uint32, peep->id);
+        utf8* curName = gCommonStringFormatBuffer;
```
Comment (janisozaur) [inline src/openrct2/actions/GuestSetNameAction.hpp]: `utf8 * curName`

**Comment by janisozaur** [comment]
I've pointed out few minor things, make sure they are applied across this whole PR.

The other thing I'm not a fan of is why do we need two have two separate actions that do almost exactly the same? Now would be a good time to clean it up a bit.

Here's the diff between staff and guest:

```diff
diff --git a/src/openrct2/actions/GuestSetNameAction.hpp b/src/openrct2/actions/GuestSetNameAction.hpp
index 2cd246138..90dc024ae 100644
--- a/src/openrct2/actions/GuestSetNameAction.hpp
+++ b/src/openrct2/actions/GuestSetNameAction.hpp
@@ -21,21 +21,22 @@
 #include "GameAction.h"
 
 #include "../cheats.h"
+#include "../peep/staff.h"
 #include "../interface/window.h"
 #include "../localisation/localisation.h"
 #include "../world/park.h"
 
-struct GuestSetNameAction : public GameActionBase<GAME_COMMAND_SET_GUEST_NAME, GameActionResult>
+struct StaffSetNameAction : public GameActionBase<GAME_COMMAND_SET_STAFF_NAME, GameActionResult>
 {
 private:
     uint16 _spriteIndex;
     std::string _name;
 
 public:
-    GuestSetNameAction() {}
-    GuestSetNameAction(uint16 spriteIndex, const std::string& name)
-        : _spriteIndex(spriteIndex),
-        _name(name)
+    StaffSetNameAction() {}
+    StaffSetNameAction(uint16 spriteIndex, const std::string& name)
+            : _spriteIndex(spriteIndex),
+              _name(name)
     {
     }
 
@@ -60,11 +61,11 @@ public:
 
         if (_name.empty())
         {
-            return std::make_unique<GameActionResult>(GA_ERROR::INVALID_PARAMETERS, STR_ERR_INVALID_NAME_FOR_GUEST);
+            return std::make_unique<GameActionResult>(GA_ERROR::INVALID_PARAMETERS, STR_STAFF_ERROR_CANT_NAME_STAFF_MEMBER);
         }
 
         rct_peep *peep = GET_PEEP(_spriteIndex);
-        if (peep->type != PEEP_TYPE_GUEST)
+        if (peep->type != PEEP_TYPE_STAFF)
         {
             log_warning("Invalid game command for sprite %u", _spriteIndex);
             return std::make_unique<GameActionResult>(GA_ERROR::INVALID_PARAMETERS, STR_NONE);
@@ -86,7 +87,7 @@ public:
         rct_string_id newUserStringId = user_string_allocate(USER_STRING_HIGH_ID_NUMBER | USER_STRING_DUPLICATION_PERMITTED, _name.c_str());
 
         rct_peep *peep = GET_PEEP(_spriteIndex);
-        if (peep->type != PEEP_TYPE_GUEST)
+        if (peep->type != PEEP_TYPE_STAFF)
         {
             log_warning("Invalid game command for sprite %u", _spriteIndex);
             return std::make_unique<GameActionResult>(GA_ERROR::INVALID_PARAMETERS, STR_NONE);
@@ -111,8 +112,8 @@ public:
 
         gfx_invalidate_screen();
 
-        // Force guest list window refresh
-        rct_window *w = window_find_by_class(WC_GUEST_LIST);
+        // Force staff list window refresh
+        rct_window *w = window_find_by_class(WC_STAFF_LIST);
         if (w != NULL)
         {
             w->no_list_items = 0;
```

How about we change the action to be refer to `peep` rather than `guest`/`staff` and infer the type of peep inside the action? What do you think @IntelOrca ?

**Comment by oli414** [comment]
Thanks for the review. I do agree on combining the two. Already noticed it but didn't wanted to be the one making the choice. How would we map it against the `GAME_COMMAND` enum in that case?

**Comment by IntelOrca** [comment]
@janisozaur sounds good

**Comment by janisozaur** [comment]
`peep` is the generic term, a `peep` can be either `guest` or `staff`, so maybe we can reuse one of the game commands and rename it to `peep` and provide a dummy one for the other? Or just remove the other altogether.

This may affect permissions, however.

**Comment by janisozaur** [comment]
@oli414 I will merge it now and I want to merge #5053 into `develop` afterwards. You can submit the unification PR later on.

**Pull request merged**